File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.cpp
class Solution {
public:
    vector<int> findThePrefixCommonArray(vector<int>& A, vector<int>& B) {
        int asz = A.size(), bsz = B.size(), msz = max(bsz, asz), cnt;
        vector<int> ans(msz, 0);
        unordered_map<int, int> amp, bmp;

        for(int i=0; i<msz; i++){
            int a = (i < asz ? A[i] : INT_MAX), b = (i < bsz ? B[i] : INT_MAX);
            if(a != INT_MAX) amp[a]++;
            if(b != INT_MAX) bmp[b]++;
            cnt = 0;
            for(auto j : amp){
                if(bmp.find(j.first) != bmp.end()){
                    cnt += min(j.second, bmp[j.first]);
                }
            }
            ans[i] = cnt;
        }

        return ans;
    }
};

File: 2766-find-the-prefix-common-array-of-two-arrays/find-the-prefix-common-array-of-two-arrays.java
class Solution {
    public int[] findThePrefixCommonArray(int[] A, int[] B) {
        int sz = A.length, ans[] = new int[sz];
        Set<Integer> one = new HashSet<>(), two = new HashSet<>();

        for(int i=0; i<sz; i++){
            if(one.contains(A[i])){
                two.add(A[i]);
            }
            else one.add(A[i]);
            if(one.contains(B[i])){
                two.add(B[i]);
            }
            else one.add(B[i]);
            ans[i] = two.size();
        }

        return ans;
    }
}